ui/src/menu.tsx

```diff
@@ -0,0 +1,45 @@
+import {
+  default as BaseMenu,
+  MenuProps as BaseMenuProps,
+} from "@mui/material/Menu";
+import {
+  createElement,
+  MouseEvent,
+  ReactElement,
+  ReactNode,
+  useCallback,
+  useState,
+} from "react";
+
+interface MenuProps
+  extends Omit<BaseMenuProps, "anchorEl" | "open" | "onClick" | "onClose"> {
+  items: ReactNode;
+}
+
+export function useMenu(
+  props: MenuProps
+): [ReactElement, (e: MouseEvent<HTMLElement | undefined>) => void] {
+  const [anchorEl, setAnchorEl] = useState<HTMLElement | undefined>();
+  const show = useCallback((e) => setAnchorEl(e.currentTarget), []);
+
+  const menu = useCallback(
+    (props: MenuProps): ReactElement => {
+      return (
+        <>
+          <BaseMenu
+            {...props}
+            anchorEl={anchorEl}
+            open={!!anchorEl}
+            onClick={() => setAnchorEl(undefined)}
+            onClose={() => setAnchorEl(undefined)}
+          >
+            {props.items}
+          </BaseMenu>
+        </>
+      );
+    },
+    [props]
+  );
+
+  return [createElement(menu, props, null), show];
```

Comment:
im trying to understand this, but not quite following based on the docs, which have the first param being a type (but you're padding a callback)?

mainly curious when the callback would be invoked, and how its not run each time.